### ISSUE TITLE
'colocate_gradients_with_ops' colocate with unused ops

### ISSUE THREAD

**Issue by ppwwyyxx** [tensorflow/tensorflow#17500]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**:
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: ubuntu 16.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**:v1.6.0-0-gd2e24b6039 1.6.0 
- **Python version**: 3.6
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: 9.0/7
- **GPU model and memory**: P100
- **Exact command to reproduce**: here

```python
import tensorflow as tf
import time

G = tf.get_default_graph()

@tf.RegisterGradient("SummaryGrad")
def summarize_grad_fn(op, x):
    my_name_is_not_x = tf.Print(x, [x])
    return x

def layer(x):
    l = tf.layers.conv2d(x, 64, 3, data_format='channels_first')
    l = tf.nn.relu(l)
    with G.gradient_override_map({"Identity": "SummaryGrad"}):
        return tf.identity(l)

img = tf.random_normal((64,3, 224,224))
l = layer(img)
l = layer(l)
l = layer(l)

loss = tf.reduce_mean(l)

opt = tf.train.GradientDescentOptimizer(0.1)
grads = opt.compute_gradients(loss, colocate_gradients_with_ops=True)
op = opt.apply_gradients(grads)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    for k in range(20):
        start = time.time()
        sess.run(op)
        if k > 2:
            print(time.time() - start)
```

If I use `colocate_gradients_with_ops=False`, or remove the `tf.Print` line, the above code runs with expected speed.
Otherwise, it places the gradients on CPUs, resulting in 2x slow down, and a lot of H2D/D2H copy shown in profiling.
Perhaps this colocate option can be made smarter.

In this case, Print op depends on some gradients, but no other gradients depend on the Print op. i.e., the Print op __does not appear on the subgraph which `grads` depends on__. Therefore the colocation seems totally unnecessary here.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Exact command to reproduce

**Comment by drpngx** [comment]
I see. I think it is working as expected, though perhaps not the behavior you would want.

I assume that you distilled the repro from a more interesting use case, maybe you can tell us a bit more about the context? I'm assuming that there is something that prevents you from using `with tf.device`?

@agarwal-ashish WDYT?

**Comment by ppwwyyxx** [comment]
The colocation constraint happens inside `tf.gradients` so I could not change this behavior from outside as long as I enable this option. I enable the option because it sometimes makes things slightly faster and usually not slower.

I may be wrong about what this option is meant to do, but I assume it's meant to make `grads = tf.gradients(..)` faster. But given that `grads` does not depend on the Print op, there is no reason to colocate with the Print op. 

Normally when `tf.gradients` is called, no unused ops like this one will be added so the behavior will always make sense, unless one uses `gradien_override_map`.
My use case is to occasionally run summaries or print ops on the gradients.

**Comment by agarwal-ashish** [comment]
colocate_gradients_with_ops is a very simple heuristic which simply tries to colocate the gradients with the original op, and may or may not be the best placement policy. Optimal placement is hard in general and has been an area of active research.

In this case, I'd suggest using tf.device or tf.colocate_with (setting it on the Identity or inside the gradient) and see if that helps.

Closing this issue since the code is working as expected.
